nova_vm/src/ecmascript/execution/environments/declarative_environment.rs

```diff
@@ -1,56 +1,292 @@
-use super::EnvironmentIndex;
-use crate::ecmascript::types::{String, Value};
-use std::{collections::HashMap, marker::PhantomData, num::NonZeroU32};
+use super::OuterEnv;
+use crate::ecmascript::{
+    execution::{agent::ExceptionType, Agent, JsResult},
+    types::{Object, Value},
+};
+use oxc_span::Atom;
+use std::collections::HashMap;
 
-#[derive(Debug, Clone, Copy)]
-pub struct DeclarativeEnvironmentIndex(NonZeroU32, PhantomData<DeclarativeEnvironment>);
-
-impl DeclarativeEnvironmentIndex {
-    pub const fn from_u32_index(value: u32) -> Self {
-        assert!(value != u32::MAX);
-        // SAFETY: Number is not max value and will not overflow to zero.
-        // This check is done manually to allow const context.
-        Self(unsafe { NonZeroU32::new_unchecked(value + 1) }, PhantomData)
-    }
-
-    pub const fn from_usize_index(value: usize) -> Self {
-        debug_assert!(value < u32::MAX as usize);
-        Self(
-            unsafe { NonZeroU32::new_unchecked(value as u32 + 1) },
-            PhantomData,
-        )
-    }
-
-    pub const fn into_index(self) -> usize {
-        self.0.get() as usize - 1
-    }
-}
-
-/// 9.1.1.1 Declarative Environment Records
-/// https://tc39.es/ecma262/#sec-declarative-environment-records
+/// ### [9.1.1.1 Declarative Environment Records](https://tc39.es/ecma262/#sec-declarative-environment-records)
+///
+/// A Declarative Environment Record is used to define the effect of ECMAScript
+/// language syntactic elements such as FunctionDeclarations,
+/// VariableDeclarations, and Catch clauses that directly associate identifier
+/// bindings with ECMAScript language values.
 #[derive(Debug)]
 pub struct DeclarativeEnvironment {
-    pub outer_env: Option<EnvironmentIndex>,
-    pub bindings: HashMap<String, Binding>,
+    /// ### \[\[OuterEnv\]\]
+    ///
+    /// See [OuterEnv].
+    outer_env: OuterEnv,
+
+    /// The environment's bindings.
+    bindings: HashMap<Atom, Binding>,
 }
 
-// TODO: Consider splitting binding this into ValueBinding
```

Comment:
**nitpick:** IMO this TODO is still relevant and not entirely covered by a bitfield packing TODO. The idea for the splitting was to (possibly) do a struct-of-vectors type of thing so that `get_binding_value` and any iteration over the bindings (GC) would not need to load the metadata.